ci_group/revolve2/ci_group/morphological_novelty_metric/_morphological_novelty_metric.py

```diff
@@ -0,0 +1,139 @@
+from math import atan2, pi, sqrt
+
+import numpy as np
+from numpy.typing import NDArray
+
+from revolve2.modular_robot import ModularRobot
+
+from ._coordinate_operations import CoordinateOperations
+from .calculate_novelty import calculate_novelty
+
+
+@dataclass(init=False)
+class MorphologicalNoveltyMetric:
+    """
+    Calculate the Morphological Novelty Score for a Population.
+
+    This metric for Morphological Novelty considers robots a distribution in space, which can be reshaped into any other distribution.
+    The work that has to be done to reshape distribution 1 to distribution 2 is used for the final novelty calculation.
+
+    A detailed description of the Algorithm can be found in:
+    Oliver Weissl, and A.E. Eiben. "Morphological-Novelty in Modular Robot Evolution". 2023 IEEE Symposium Series on Computational Intelligence (SSCI)(pp. 1066-1071). IEEE, 2023.
+    """
+
+    _coordinates: list[NDArray[np.float128]]
+    _magnitudes: list[list[float]]
+    _orientations: list[list[tuple[float, float]]]
+    _histograms: NDArray[np.float128]
+    _int_histograms: NDArray[np.int64]
+    _novelty_scores: NDArray[np.float64]
+
+    _NUM_BINS: int = 20
+    """The amount of bins in the histogram. Increasing this allows for more detail, but risks sparseness, while lower values generalize more."""
+    _INT_CASTER: int = 10_000
+    """Casting floats to INT allows to mitigate floating-point issues in the distribution reshaping. The higher the number, the more presicion you get."""
+
+    def get_novelty_from_population(
+        self,
+        population: list[ModularRobot],
+        cob_heuristic: bool = False,
+    ) -> list[float]:
+        """
+        Get the morphological novelty score for individuals in a population.
+
+        :param population: The population of robots.
+        :param cob_heuristic: Whether the heuristic approximation for change of basis is used.
+        :return: The novelty scores.
+        """
+        instances = len(population)
+        bodies = [robot.body for robot in population]
+
+        coord_ops = CoordinateOperations()
+        self._coordinates = coord_ops.coords_from_bodies(
+            bodies, cob_heuristics=cob_heuristic
+        )
+
+        self._histograms = np.empty(
+            shape=(instances, self._NUM_BINS, self._NUM_BINS), dtype=np.float64
+        )
+        self._int_histograms = np.empty(
+            shape=(instances, self._NUM_BINS, self._NUM_BINS), dtype=np.int64
+        )
+
+        self._coordinates_to_magnitudes_orientation()
+        self._gen_gradient_histogram()
+        self._normalize_cast_int()
+
+        self._novelty_scores = calculate_novelty(
+            self._int_histograms, self._int_histograms.shape[0], self._NUM_BINS
+        )
+        max_novelty = self._novelty_scores.max()
+        try:
+            return [float(score / max_novelty) for score in self._novelty_scores]
+        finally:
+            self.__reset_class()
```

Comment:
See feedback at _coordinate_operations.py. Using try finally like this is just not common habit